apis/apiextensions/v1/composition_transforms.go

```diff
@@ -221,15 +387,28 @@ type StringTransformRegexp struct {
 	Group *int `json:"group,omitempty"`
 }
 
-// The list of supported ConvertTransform input and output types.
+// TransformIOType defines the type of a ConvertTransform.
+type TransformIOType string
+
+// The list of supported Transform input and output types.
 const (
-	ConvertTransformTypeString  = "string"
-	ConvertTransformTypeBool    = "bool"
-	ConvertTransformTypeInt     = "int"
-	ConvertTransformTypeInt64   = "int64"
-	ConvertTransformTypeFloat64 = "float64"
+	TransformIOTypeString  TransformIOType = "string"
+	TransformIOTypeBool    TransformIOType = "bool"
+	TransformIOTypeInt     TransformIOType = "int"
+	TransformIOTypeInt64   TransformIOType = "int64"
+	TransformIOTypeFloat64 TransformIOType = "float64"
 )
 
+// IsValid checks if the given TransformIOType is valid.
+func (c TransformIOType) IsValid() bool {
```

Comment:
The validations we are doing for enum values are already enforced by Kube API Server via [the schema](https://github.com/crossplane/crossplane/blob/33c8a2c52f6465513a3e1873613446b307e3356b/cluster/crds/apiextensions.crossplane.io_compositions.yaml#L246C37-L258). 

In the case of admission webhook validation, I believe we will never hit a failing validation in this function but this could be helpful for external tools like linters that rely on this functionality. However, I am still not sure if it worths the maintenance burden in the long run, e.g. whenever you add a new value to enum, don't forget to update the `IsValid` method. Especially given that we don't have a concrete use case yet.

I would rather prefer reusing the existing mechanism rather than having it in multiple places. For example, could there be a way to run validation of an object against its [open API v3 schema](https://github.com/crossplane/crossplane/blob/33c8a2c52f6465513a3e1873613446b307e3356b/cluster/crds/apiextensions.crossplane.io_compositions.yaml#L31) and get validations like this out of the box in the use cases other than admission webhook mode?

Reply:
> The missing case is going to be caught by the exhaustive linter e.g. removing TransformIOTypeFloat64 from the switch and running golangci-lint

Good call, so maintenance may not be a big concern. 

However, I would still vote for not duplicating the validations that will be assured by the schema for the sake of leanness. With Kube API server validating them, there is no possibility for it to happen at runtime, e.g. with [this](https://github.com/crossplane/crossplane/blob/4cf4134d03771ec660aee3770257d125687d3b84/cluster/crds/apiextensions.crossplane.io_compositions.yaml#L394-L401).

Reply:
👍 agreed, removing them

Reply:
removed and changed a few things accordingly: https://github.com/crossplane/crossplane/pull/3921/commits/8681c201751761c07d0559f3d3c1875f2832c0c4